From c72381a56053b9a6dd85dda36dd7befdc4a6c51f Mon Sep 17 00:00:00 2001
From: Pavel Nakonechnyi <zorg@altlinux.org>
Date: Tue, 9 Jul 2019 21:35:30 +0200
Subject: [PATCH 5/6] pico-imx8mq: enable 'gpt' and 'part' commands

---
 include/configs/pico-imx8mq.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/configs/pico-imx8mq.h b/include/configs/pico-imx8mq.h
index cd23836089..d676df3a4b 100644
--- a/include/configs/pico-imx8mq.h
+++ b/include/configs/pico-imx8mq.h
@@ -336,6 +336,11 @@
 #define CONFIG_IMX_VIDEO_SKIP
 #endif
 
+#ifndef CONFIG_SPL_BUILD
+#define CONFIG_CMD_GPT
+#define CONFIG_CMD_PART
+#endif
+
 #if defined(CONFIG_ANDROID_SUPPORT)
 #include "pico-imx8mq_android.h"
 #elif defined (CONFIG_ANDROID_THINGS_SUPPORT)
-- 
2.21.0

