--- a/include/configs/pico-imx8mq.h
+++ b/include/configs/pico-imx8mq.h
@@ -124,6 +124,15 @@
 #define SYS_AUXCORE_BOOTDATA_DDR	0x80000000
 #define SYS_AUXCORE_BOOTDATA_TCM	0x007E0000
 
+/* trick to prepare MMC partitions */
+#define CONFIG_BOOTCOMMAND "" \
+	"if test ${part_resized} = 0; then " \
+		"run setup_emmc;" \
+		"setenv part_resized 1; saveenv;" \
+		"reset;" \
+	"fi;" \
+	"run distro_bootcmd"
+
 /* Initial environment variables */
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	CONFIG_MFG_ENV_SETTINGS \
@@ -146,6 +155,12 @@
 	"bootenv=uEnv.txt\0" \
 	"rescuefile=tnrescue.itb\0" \
 	"fit_addr=0x45800000\0" \
+	"rootfs_start=8MiB\0" \
+	"rootfs_size=7400MiB\0" \
+	"finduuidrootfs=part uuid mmc 0:1 uuid_gpt_rootfs\0" \
+	"partitions=uuid_disk=${uuid_gpt_disk};name=rootfs,start=${rootfs_start},size=${rootfs_size},uuid=${uuid_gpt_rootfs}\0" \
+	"part_resized=0\0" \
+	"setup_emmc=run finduuidrootfs;gpt guid mmc 0 uuid_gpt_disk;gpt write mmc 0 $partitions;\0" \
 	BOOTENV
 
 #define BOOT_TARGET_DEVICES(func) \
